ddtrace/internal/telemetry/writer.py

```diff
@@ -588,15 +546,18 @@ def _app_client_configuration_changed_event(self, configurations):
         }
         self.add_event(payload, "app-client-configuration-change")
 
-    def _update_dependencies_event(self, newly_imported_deps: List[str]):
+    def _update_dependencies_event(self):
         """Adds events to report imports done since the last periodic run"""
 
         if not config._telemetry_dependency_collection or not self._enabled:
             return
 
         with self._lock:
-            packages = update_imported_dependencies(self._imported_dependencies, newly_imported_deps)
+            latest_modules = set(sys.modules.keys())
+            new_modules = latest_modules - self._imported_dependencies
```

Comment:
Is this faster than the module watchdog solution when `sys.modules` is a large collection? Do we have some numbers?

Reply:
I don't have any numbers for this, but easy enough to measure the set creating and diffing overhead with a large dataset.

What do you think is a realistic size of sys.modules to test with? 1000 entries? 10,000 entries? I would think anything higher than that would be crazy.


Also, the part of the module watchdog approach which was slow was going from module import -> file -> file origin -> distribution. This was the main source of overhead the current implementation. As long as performance overhead (when I shared numbers) is acceptable, I'd greatly prefer this approach to having the additional moving pieces of a module watchdog.

Reply:
```python
import random
import string
import pyperf


def random_string():
    return "".join(random.choice(string.ascii_letters) for _ in range(10))


def bench_module_diff(seen_modules, sys_modules):
    latest_modules = set(sys_modules.keys())
    _ = latest_modules - seen_modules


runner = pyperf.Runner()

names = [random_string() for _ in range(10_000)]

for num_existing in [0, 500, 2_000, 5_000, 10_000]:
    for num_modules in [500, 2_000, 5_000, 10_000]:
        if num_modules < num_existing:
            continue

        runner.bench_func(
            f"module_diff[{num_existing},{num_modules}]",
            bench_module_diff,
            set(names[:num_existing]),
            {names[i]: True for i in range(num_modules)},
        )
```

The results are `module_diff[{num modules names already "seen"}, {num modules in sys.modules}]`


```
.....................
module_diff[0,500]: Mean +- std dev: 9.22 us +- 0.59 us
.....................
module_diff[0,2000]: Mean +- std dev: 65.7 us +- 3.9 us
.....................
module_diff[0,5000]: Mean +- std dev: 233 us +- 6 us
.....................
module_diff[0,10000]: Mean +- std dev: 378 us +- 6 us
.....................
module_diff[500,500]: Mean +- std dev: 9.67 us +- 0.64 us
.....................
module_diff[500,2000]: Mean +- std dev: 94.8 us +- 3.4 us
.....................
module_diff[500,5000]: Mean +- std dev: 232 us +- 5 us
.....................
module_diff[500,10000]: Mean +- std dev: 378 us +- 6 us
.....................
module_diff[2000,2000]: Mean +- std dev: 63.5 us +- 3.1 us
.....................
module_diff[2000,5000]: Mean +- std dev: 277 us +- 5 us
.....................
module_diff[2000,10000]: Mean +- std dev: 388 us +- 7 us
.....................
module_diff[5000,5000]: Mean +- std dev: 207 us +- 4 us
.....................
module_diff[5000,10000]: Mean +- std dev: 521 us +- 17 us
.....................
module_diff[10000,10000]: Mean +- std dev: 367 us +- 9 us
```


I think something between the "500" and "2000" cases are the most realistic for most applications? So somewhere between 0.01ms and 0.1ms per-operation to get the list of newly imported modules.